README
========================
Python Client for Riak
========================


Documentation
==============

`Documentation for the Riak Python Client Library <http://basho.github.com/riak-python-client/index.html>`_ is available here. 
The documentation source is found in `docs/ subdirectory 
<https://github.com/basho/riak-python-client/tree/master/docs>`_ and can be 
built with `Sphinx <http://sphinx.pocoo.org/>`_.

Documentation for Riak is available at http://wiki.basho.com/display/RIAK/Riak

Install
=======

You must have `Protocol Buffers`_ installed before you can install the Riak Client. From the Riak Python Client root directory, execute::

    python setup.py install

Unit Test
===========
To run the unit tests, execute::

    python setup.py test


This tutorial assumes basic working knowledge of how Riak works & what it can
do. If you need a more comprehensive overview how to use Riak, please check out
the `Riak Fast Track`_.

.. _`Riak Fast Track`: http://wiki.basho.com/display/RIAK/The+Riak+Fast+Track

========================
Tutorial
========================

Quick Start
===========

For the impatient, simple usage of the official Python binding for Riak looks
like::

    import riak

    # Connect to Riak.
    client = riak.RiakClient()

    # Choose the bucket to store data in.
    bucket = client.bucket('test')


    # Supply a key to store data under.
    # The ``data`` can be any data Python's ``json`` encoder can handle.
    person = bucket.new('riak_developer_1', data={
        'name': 'John Smith',
        'age': 28,
        'company': 'Mr. Startup!',
    })
    # Save the object to Riak.
    person.store()


Connecting To Riak
==================

There are two supported ways to connect to Riak, the HTTP interface & the
`Protocol Buffers`_ interface. Both provide the same API & full access to
Riak.

The HTTP interface is easier to setup & is well suited for development use. It
is the slower of the two interfaces, but if you are only making a handful of
requests, it is more than capable.

The Protocol Buffers (also called ``protobuf``) is more difficult to setup but
is significantly faster (2-3x) and is more suitable for production use. This
interface is better suited to a higher number of requests.

.. _`Protocol Buffers`: http://code.google.com/p/protobuf/

To use the HTTP interface and connecting to a local Riak on the default port,
no arguments are needed::

    import riak

    client = riak.RiakClient()

The constructor also configuration options such as ``host``, ``port`` &
``prefix``. Please refer to the `client` documentation for full details.

To use the Protocol Buffers interface::

    import riak

    client = riak.RiakClient(port=8087, transport_class=riak.RiakPbcTransport)

.. warning:

  Riak's default port is 8098. However, when using the Protocol Buffers, the
  Riak listens on port 8087. If you forget this, you will *NOT* get an
  immediate error, but will instead receive an error when fetching or storing
  data to the effect of ``RiakError: 'Socket returned short read 135 -
  expected 8192'``.

The ``transport_class`` argument indicates to the client which backend to use.
We didn't need to specify it in the HTTP example because
``riak.RiakHttpTransport`` is the default class.


Using Buckets
=============

Buckets in Riak's terminology are segmented keyspaces. They are a way to
categorize different types of data and are roughly analogous to tables in an
RDBMS.

Once you have a ``client``, selecting a bucket is simple. Provide a string of
the name of the bucket to use::

    test_bucket = client.bucket('test')

If the bucket does not exist, Riak will create it for you. You can also open
as many buckets as you need::

    user_bucket = client.bucket('user')
    profile_bucket = client.bucket('profile')
    status_bucket = client.bucket('status')

If needed, you can also manually instantiate a bucket like so::

    user_bucket = riak.RiakBucket(client, 'user')

The buckets themselves provide many different methods. The most commonly used
are:

* ``get`` - Fetches a key's value (decoded from JSON).
* ``get_binary`` - Also fetches a key's raw value (plain text or binary).
* ``new`` - Creates a new key/value pair (encoded in JSON).
* ``new_binary`` - Creates a new key/raw value pair.

See the full `bucket` documentation for the other methods.


Storing Keys/Values
===================

Once you've got a working client/bucket, the next task at hand is storing data.
Riak provides several ways to store your data, but the most common are a
JSON-encoded structure or a binary blob.

To store JSON-encoded data, you'd do something like the following::

  import riak

  client = riak.RiakClient()
  user_bucket = client.bucket('user')

  # We're creating the user data & keying off their username.
  new_user = user_bucket.new('johndoe', data={
      'first_name': 'John',
      'last_name': 'Doe',
      'gender': 'm',
      'website': 'http://example.com/',
      'is_active': True,
  })
  # Note that the user hasn't been stored in Riak yet.
  new_user.store()

Note that any data Python's ``json`` (or ``simplejson``) encoder can handle is
fair game.

As mentioned, Riak can also handle binary data, such as images, audio files,
etc. Storing binary data looks almost identical::

  import riak

  client = riak.RiakClient()
  user_photo_bucket = client.bucket('user_photo')

  # For example purposes, we'll read a file off the filesystem, but you can get
  # the data from anywhere.
  the_photo_data = open('/tmp/johndoe_headshot.jpg', 'rb').read()

  # We're storing the photo in a different bucket but keyed off the same
  # username.
  new_user = user_bucket.new_binary('johndoe', data=the_photo_data, content_type='image/jpeg')
  new_user.store()

You can also manually store data by using ``RiakObject``::

  import riak
  import time
  import uuid

  client = riak.RiakClient()
  status_bucket = client.bucket('status')

  # We use ``uuid.uuid1()`` here to create a unique identifier for the status.
  new_status = RiakObject(client, status_bucket, uuid.uuid1())

  # Add in the data you want to store.
  new_status.set_data({
      'message': 'First post!',
      'created': time.time(),
      'is_public': True,
  })

  # Set the content type.
  new_status.set_content_type('application/json')

  # We want to do JSON-encoding on the value.
  new_status._encode_data = True

  # Again, make sure you save it.
  new_status.store()


Getting Single Values Out
=========================

Storing data is all well and good, but you'll need to get that data out at a
later date.

Riak provides several ways to get data out, though fetching single key/value
pairs is the easiest. Just like storing the data, you can pull the data out
in either the JSON-decoded form or a binary blob. Getting the JSON-decoded
data out looks like::

  import riak

  client = riak.RiakClient()
  user_bucket = client.bucket('user')

  johndoe = user_bucket.get('johndoe')

  # You've now got a ``RiakObject``. To get at the values in a dictionary
  # form, call:
  johndoe_dict = johndoe.get_data()

Getting binary data out looks like::

  import riak

  client = riak.RiakClient()
  user_photo_bucket = client.bucket('user_photo')

  johndoe = user_photo_bucket.get_binary('johndoe')

  # You've now got a ``RiakObject``. To get at the binary data, call:
  johndoe_headshot = johndoe.get_data()

Manually fetching data is also possible::

  import riak

  client = riak.RiakClient()
  status_bucket = client.bucket('status')

  # We're using the UUID generated from the above section.
  first_post_status = RiakObject(client, status_bucket, '39fbee54-fb82-11df-a2cf-d49a20c04e6a')
  first_post_status._encode_data = True
  r = status_bucket.get_r(r)

  # Calling ``reload`` will cause the ``RiakObject`` instance to load fresh
  # data/metadata from Riak.
  first_post_status.reload(r)

  # Finally, pull out the data.
  message = first_post_status.get_data()['message']


Fetching Data Via Map/Reduce
============================

When you need to work with larger sets of data, one of the tools at your
disposal is MapReduce_. This technique iterates over all of the data, returning
data from the map phase & combining all the different maps in the reduce
phase(s).

.. _MapReduce: http://wiki.basho.com/display/RIAK/MapReduce

To perform a map operation, such as returning all active users, you can do
something like::

  import riak

  client = riak.RiakClient()
  # First, you need to ``add`` the bucket you want to MapReduce on.
  query = client.add('user')

RELEASE_NOTES.md
# Riak Python Client Release Notes

## 2.4.1 Patch Release - 2016-02-03

* [Riak TS: Millisecond precision](https://github.com/basho/riak-python-client/issues/430)
* [Fix release process](https://github.com/basho/riak-python-client/issues/429)

## 2.4.0 Feature Release - 2016-01-13

This release enhances Riak Time Series functionality.

* [Encapsulate table description](https://github.com/basho/riak-python-client/pull/422)

## 2.3.0 Feature Release - 2015-12-14

Release 2.3.0 features support for new
[time series](https://github.com/basho/riak-python-client/pull/416)
functionality.

This is release retires support for Python 2.6.x but adds support for
Python 3.5.x.

There are also many bugfixes and new enhancements:

* [Protocol buffers are now integrated into the Python Client]
  (https://github.com/basho/riak-python-client/pull/418)
* [Support for Preflists and Write-Once bucket types]
  (https://github.com/basho/riak-python-client/pull/414)
* [Support Riak 2.1.1]
  (https://github.com/basho/riak-python-client/pull/407)
* [Native SSL support for Python 2.7.9+]
  (https://github.com/basho/riak-python-client/pull/397)


## 2.2.0 Feature Release - 2014-12-18

Release 2.2.0 features support for
[Python 3](https://github.com/basho/riak-python-client/pull/379),
specifically 3.3 and 3.4.  This version uses the native SSL security instead
of [pyOpenSSL](http://pypi.python.org/pypi/pyOpenSSL) which is required
for the Python 2 series.

This release also includes many bugfixes and enhancements, most
notably:

* [Fixed an issue with the implementation of `Mapping.__iter__`]
  (https://github.com/basho/riak-python-client/pull/367)
* [Test client certificate generation updated]
  (https://github.com/basho/riak-python-client/pull/373)
* [Protocol Buffers had a socket.send issue]
  (https://github.com/basho/riak-python-client/pull/382)
* [Support for bucket types in Map/Reduce jobs added]
  (https://github.com/basho/riak-python-client/pull/385)
* [Race condition in `RiakBucket` creation fixed]
  (https://github.com/basho/riak-python-client/pull/386)
* [Data Types can now be deleted]
  (https://github.com/basho/riak-python-client/pull/387)
* [2i Range Queries with a zero end index now work]
  (https://github.com/basho/riak-python-client/pull/388)


## 2.1.0 Feature Release - 2014-09-03

Release 2.1.0 features support for Riak 2.0 capabilities including:

* Bucket Types
* Riak Data Types (CRDTs)
* Search 2.0 (codename Yokozuna)
* Security: SSL/TLS, Authentication, and Authorization

As a result of the new security features, the package now depends on
[pyOpenSSL](http://pypi.python.org/pypi/pyOpenSSL) and will warn if
your version of OpenSSL is too old.

This release also includes many bugfixes and enhancements, most
notably:

* The default protocol is now 'pbc', not 'http'.
* When used correctly, streaming requests no longer result in leaks
  from the connection pool.
* The size of the multiget worker pool can be set when initializing
  the client.
* Secondary index queries can now iterate over all pages in a query.
* The number of times a request is retried after network failure is
  now configurable.
* The additional request options `basic_quorum` and `notfound_ok` are
  now supported.

## 2.0.3 Patch Release - 2014-03-06

Release 2.0.3 includes support for 1.4.4's 2I regexp feature and fixes
a few bugs:

* Docs generation now uses the version from the top-level package.
* Some internal uses of the deprecated RiakClient.solr were removed.
* More errors will be caught and propagated properly from multiget
  requests, preventing deadlocks on the caller side.

## 2.0.2 Patch release - 2013-11-18

Release 2.0.2 includes support for the 1.4.1+ "timeout" option on
secondary index queries.

## 2.0.1 Patch release - 2013-08-28

Release 2.0.1 includes a minor compatibility fix for Python 2.6 and an
updated README.

## 2.0.0 Feature Release - 2013-07-30

Release 2.0 is the culmination of many months of rearchitecting the
client. Highlights:

* Automatic connection to multiple nodes, with request retries,
  through a thread-safe connection pool.
* All Riak 1.3 and 1.4 features, including bucket properties,
  paginating and streaming secondary indexes, CRDT counters,
  client-specified timeouts, and more.

commands.py
import csv
import os
import os.path
import re
import shutil

from distutils.core import Command
from distutils.errors import DistutilsOptionError
from distutils.file_util import write_file
from distutils import log
from string import Template
from subprocess import Popen, PIPE


__all__ = ['build_messages',
           'setup_security', 'enable_security', 'disable_security',
           'setup_timeseries',
           'preconfigure', 'configure']


# Exception classes used by this module.
class CalledProcessError(Exception):
    """This exception is raised when a process run by check_call() or
    check_output() returns a non-zero exit status.
    The exit status will be stored in the returncode attribute;
    check_output() will also store the output in the output attribute.
    """
    def __init__(self, returncode, cmd, output=None):
        self.returncode = returncode
        self.cmd = cmd
        self.output = output

    def __str__(self):
        return "Command '%s' returned non-zero exit status %d" % (self.cmd,
                                                                  self
                                                                  .returncode)


def check_output(*popenargs, **kwargs):
    """Run command with arguments and return its output as a byte string.

    If the exit code was non-zero it raises a CalledProcessError.  The
    CalledProcessError object will have the return code in the returncode
    attribute and output in the output attribute.

    The arguments are the same as for the Popen constructor.  Example:

    >>> check_output(["ls", "-l", "/dev/null"])
    'crw-rw-rw- 1 root root 1, 3 Oct 18  2007 /dev/null\n'

    The stdout argument is not allowed as it is used internally.
    To capture standard error in the result, use stderr=STDOUT.

    >>> import sys
    >>> check_output(["/bin/sh", "-c",
    ...               "ls -l non_existent_file ; exit 0"],
    ...              stderr=sys.stdout)
    'ls: non_existent_file: No such file or directory\n'
    """
    if 'stdout' in kwargs:
        raise ValueError('stdout argument not allowed, it will be '
                         'overridden.')
    process = Popen(stdout=PIPE, *popenargs, **kwargs)
    output, unused_err = process.communicate()
    retcode = process.poll()
    if retcode:
        cmd = kwargs.get("args")
        if cmd is None:
            cmd = popenargs[0]
        raise CalledProcessError(retcode, cmd, output=output)
    return output

try:
    import simplejson as json
except ImportError:
    import json


class bucket_type_commands:
    def initialize_options(self):
        self.riak_admin = None

    def finalize_options(self):
        if self.riak_admin is None:
            raise DistutilsOptionError("riak-admin option not set")

    def run(self):
        if self._check_available():
            for name in self._props:
                self._create_and_activate_type(name, self._props[name])

    def check_output(self, *args, **kwargs):
        if self.dry_run:
            log.info(' '.join(args))
            return bytearray()
        else:
            return check_output(*args, **kwargs)

    def _check_available(self):
        try:
            self.check_btype_command("list")
            return True
        except CalledProcessError:
            log.error("Bucket types are not supported on this Riak node!")
            return False

    def _create_and_activate_type(self, name, props):
        # Check status of bucket-type
        exists = False
        active = False
        try:
            status = self.check_btype_command('status', name)
        except CalledProcessError as e:
            status = e.output

        exists = ('not an existing bucket type' not in status.decode('ascii'))
        active = ('is active' in status.decode('ascii'))

        if exists or active:
            log.info("Updating {0} bucket-type with props {1}"
                     .format(repr(name), repr(props)))
            self.check_btype_command("update", name,
                                     json.dumps({'props': props},
                                                separators=(',', ':')))
        else:
            log.info("Creating {0} bucket-type with props {1}"
                     .format(repr(name), repr(props)))
            self.check_btype_command("create", name,
                                     json.dumps({'props': props},
                                                separators=(',', ':')))

        if not active:
            log.info('Activating {0} bucket-type'.format(repr(name)))
            self.check_btype_command("activate", name)

    def check_btype_command(self, *args):
        cmd = self._btype_command(*args)
        return self.check_output(cmd)

    def run_btype_command(self, *args):
        self.spawn(self._btype_command(*args))

    def _btype_command(self, *args):
        cmd = [self.riak_admin, "bucket-type"]
        cmd.extend(args)
        return cmd


class setup_timeseries(bucket_type_commands, Command):
    """
    Creates bucket-types appropriate for timeseries.
    """

    description = "create bucket-types used in timeseries tests"

    user_options = [
        ('riak-admin=', None, 'path to the riak-admin script')
    ]

    _props = {
        'GeoCheckin': {
            'n_val': 3,
            'table_def': '''
                CREATE TABLE GeoCheckin (
                    geohash varchar not null,
                    user varchar not null,
                    time timestamp not null,
                    weather varchar not null,
                    temperature double,
                    PRIMARY KEY(
                        (geohash, user, quantum(time, 15, m)),
                        geohash, user, time
                    )
                )'''
        }
    }


class security_commands(object):
    def check_security_command(self, *args):
        cmd = self._security_command(*args)
        return self.check_output(cmd)

    def run_security_command(self, *args):
        self.spawn(self._security_command(*args))

    def _security_command(self, *args):
        cmd = [self.riak_admin, "security"]
        if isinstance(args, tuple):
            for elem in args:
                cmd.extend(elem)
        else:
            cmd.extend(args)
        return cmd

    def check_output(self, *args, **kwargs):
        if self.dry_run:
            log.info(' '.join(args))
            return bytearray()
        else:
            return check_output(*args, **kwargs)


class setup_security(Command, security_commands):
    """
    Sets up security for testing. By default this will create:

    * User `testuser` with password `testpassword`
    * User `certuser` with password `certpass`
    * Two security sources
    * Permissions on
        * riak_kv.get
        * riak_kv.put
        * riak_kv.delete
        * riak_kv.index
        * riak_kv.list_keys
        * riak_kv.list_buckets
        * riak_kv.mapreduce
        * riak_core.get_bucket
        * riak_core.set_bucket
        * riak_core.get_bucket_type
        * riak_core.set_bucket_type
        * search.admin
        * search.query
    """

    description = "create security settings used in integration tests"

    user_options = [
        ('riak-admin=', None, 'path to the riak-admin script'),
        ('username=', None, 'test user account'),
        ('password=', None, 'password for test user account'),
        ('certuser=', None, 'certificate test user account'),
        ('certpass=', None, 'password for certificate test user account')
    ]

    _commands = [
        "add-user $USERNAME password=$PASSWORD",
        "add-source $USERNAME 127.0.0.1/32 password",
        "add-user $CERTUSER password=$CERTPASS",
        "add-source $CERTUSER 127.0.0.1/32 certificate"
    ]

    _grants = {
        "riak_kv.get": ["any"],
        "riak_kv.get_preflist": ["any"],
        "riak_kv.put": ["any"],
        "riak_kv.delete": ["any"],
        "riak_kv.index": ["any"],
        "riak_kv.list_keys": ["any"],
        "riak_kv.list_buckets": ["any"],
        "riak_kv.mapreduce": ["any"],
        "riak_core.get_bucket": ["any"],
        "riak_core.set_bucket": ["any"],
        "riak_core.get_bucket_type": ["any"],
        "riak_core.set_bucket_type": ["any"],
        "search.admin": ["index", "schema"],
        "search.query": ["index", "schema"]
    }

    def initialize_options(self):
        self.riak_admin = None
        self.username = None
        self.password = None
        self.certuser = None
        self.certpass = None

    def finalize_options(self):
        if self.riak_admin is None:
            raise DistutilsOptionError("riak-admin option not set")
        if self.username is None:
            self.username = 'testuser'
        if self.password is None:
            self.password = 'testpassword'
        if self.certuser is None:
            self.certuser = 'certuser'
        if self.certpass is None:
            self.certpass = 'certpass'

    def run(self):
        if self._check_available():
            for cmd in self._commands:
                # Replace the username and password if specified
                s = Template(cmd)
                newcmd = s.substitute(USERNAME=self.username,
                                      PASSWORD=self.password,
                                      CERTUSER=self.certuser,
                                      CERTPASS=self.certpass)
                log.info("Security command: {0}".format(repr(newcmd)))
                self.run_security_command(tuple(newcmd.split(' ')))
            for perm in self._grants:
                self._apply_grant(perm, self._grants[perm])

    def _check_available(self):
        try:
            self.check_security_command("status")
            return True
        except CalledProcessError:
            log.error("Security is not supported on this Riak node!")
            return False

    def _apply_grant(self, perm, targets):
        for target in targets:
            cmd = ["grant", perm, "on", target, "to", self.username]
            log.info("Granting permission {0} on {1} to {2}"
                     .format(repr(perm), repr(target), repr(self.username)))
            self.run_security_command(cmd)
            cmd = ["grant", perm, "on", target, "to", self.certuser]
            log.info("Granting permission {0} on {1} to {2}"
                     .format(repr(perm), repr(target), repr(self.certuser)))
            self.run_security_command(cmd)


class enable_security(Command, security_commands):
    """
    Actually turn on security.
    """
    description = "turn on security within Riak"

    user_options = [
        ('riak-admin=', None, 'path to the riak-admin script'),
    ]

    def initialize_options(self):
        self.riak_admin = None

    def finalize_options(self):
        if self.riak_admin is None:
            raise DistutilsOptionError("riak-admin option not set")

    def run(self):
        cmd = "enable"
        self.run_security_command(tuple(cmd.split(' ')))


class disable_security(Command, security_commands):
    """
    Actually turn off security.
    """
    description = "turn off security within Riak"

    user_options = [
        ('riak-admin=', None, 'path to the riak-admin script'),
    ]

    def initialize_options(self):
        self.riak_admin = None

    def finalize_options(self):
        if self.riak_admin is None:
            raise DistutilsOptionError("riak-admin option not set")

    def run(self):
        cmd = "disable"
        self.run_security_command(tuple(cmd.split(' ')))


class preconfigure(Command):
    """
    Sets up security configuration.

    * Update these lines in riak.conf
        * storage_backend = leveldb
        * search = on
        * listener.protobuf.internal = 127.0.0.1:8087
        * listener.http.internal = 127.0.0.1:8098
        * listener.https.internal = 127.0.0.1:18098
        * ssl.certfile = $pwd/tests/resources/server.crt
        * ssl.keyfile = $pwd/tests/resources/server.key
        * ssl.cacertfile = $pwd/tests/resources/ca.crt
        * check_crl = off
    """

    description = "preconfigure security settings used in integration tests"

    user_options = [
        ('riak-conf=', None, 'path to the riak.conf file'),
        ('host=', None, 'IP of host running Riak'),
        ('pb-port=', None, 'protocol buffers port number'),
        ('https-port=', None, 'https port number')
    ]

    def initialize_options(self):
        self.riak_conf = None
        self.host = "127.0.0.1"
        self.pb_port = "8087"
        self.http_port = "8098"
        self.https_port = "18098"

    def finalize_options(self):
        if self.riak_conf is None:
            raise DistutilsOptionError("riak-conf option not set")

    def run(self):
        self.cert_dir = os.path.dirname(os.path.realpath(__file__)) + \
            "/riak/tests/resources"
        self._update_riak_conf()

    def _update_riak_conf(self):
        http_host = self.host + ':' + self.http_port
        https_host = self.host + ':' + self.https_port
        pb_host = self.host + ':' + self.pb_port
        self._backup_file(self.riak_conf)
        conf = None
        with open(self.riak_conf, 'r', buffering=1) as f:
            conf = f.read()
        conf = re.sub(r'search\s+=\s+off', r'search = on', conf)
        conf = re.sub(r'##[ ]+ssl\.', r'ssl.', conf)
        conf = re.sub(r'ssl.certfile\s+=\s+\S+',
                      r'ssl.certfile = ' + self.cert_dir + '/server.crt',
                      conf)
        conf = re.sub(r'storage_backend\s+=\s+\S+',
                      r'storage_backend = leveldb',
                      conf)
        conf = re.sub(r'ssl.keyfile\s+=\s+\S+',
                      r'ssl.keyfile = ' + self.cert_dir + '/server.key',
                      conf)
        conf = re.sub(r'ssl.cacertfile\s+=\s+\S+',
                      r'ssl.cacertfile = ' + self.cert_dir +
                      '/ca.crt',
                      conf)
        conf = re.sub(r'#*[ ]*listener.http.internal\s+=\s+\S+',
                      r'listener.http.internal = ' + http_host,
                      conf)
        conf = re.sub(r'#*[ ]*listener.https.internal\s+=\s+\S+',
                      r'listener.https.internal = ' + https_host,
                      conf)
        conf = re.sub(r'listener.protobuf.internal\s+=\s+\S+',
                      r'listener.protobuf.internal = ' + pb_host,
                      conf)
        conf += 'check_crl = off\n'
        # Older versions of OpenSSL client library need to match on the server
        conf += 'tls_protocols.tlsv1 = on\n'
        conf += 'tls_protocols.tlsv1.1 = on\n'
        with open(self.riak_conf, 'w', buffering=1) as f:
            f.write(conf)

    def _backup_file(self, name):
        backup = name + ".bak"
        if os.path.isfile(name):
            shutil.copyfile(name, backup)
        else:
            log.info("Cannot backup missing file {0}".format(repr(name)))


class configure(Command):
    """
    Sets up security configuration.

    * Run setup_security
    """

    description = "security settings for testing"

    user_options = setup_security.user_options

    def initialize_options(self):
        self.riak_admin = None
        self.username = None
        self.password = None

    def finalize_options(self):
        security = self.distribution.get_command_obj('setup_security')
        security.riak_admin = self.riak_admin
        security.username = self.username
        security.password = self.password

    def run(self):
        # Run all relevant sub-commands.
        for cmd_name in self.get_sub_commands():
            self.run_command(cmd_name)

    sub_commands = [('setup_security', None)]


class ComparableMixin(object):
    def _compare(self, other, method):
        try:
            return method(self._cmpkey(), other._cmpkey())
        except (AttributeError, TypeError):
            # _cmpkey not implemented, or return different type,
            # so I can't compare with "other".
            return NotImplemented

    def __lt__(self, other):
        return self._compare(other, lambda s, o: s < o)

    def __le__(self, other):
        return self._compare(other, lambda s, o: s <= o)

    def __eq__(self, other):
        return self._compare(other, lambda s, o: s == o)

    def __ge__(self, other):
        return self._compare(other, lambda s, o: s >= o)

    def __gt__(self, other):
        return self._compare(other, lambda s, o: s > o)

    def __ne__(self, other):
        return self._compare(other, lambda s, o: s != o)


class MessageCodeMapping(ComparableMixin):
    def __init__(self, code, message, proto):
        self.code = int(code)
        self.message = message
        self.proto = proto
        self.message_code_name = self._message_code_name()
        self.module_name = 'riak.pb.{0}_pb2'.format(self.proto)
        self.message_class = self._message_class()

    def _cmpkey(self):
        return self.code

    def __hash__(self):
        return self.code

    def _message_code_name(self):
        strip_rpb = re.sub(r"^Rpb", "", self.message)
        word = re.sub(r"([A-Z]+)([A-Z][a-z])", r'\1_\2', strip_rpb)
        word = re.sub(r"([a-z\d])([A-Z])", r'\1_\2', word)
        word = word.replace("-", "_")
        return "MSG_CODE_" + word.upper()

    def _message_class(self):
        try:
            pbmod = __import__(self.module_name, globals(), locals(),
                               [self.message])
            klass = pbmod.__dict__[self.message]
            return klass
        except KeyError:
            log.warn("Did not find '%s' message class in module '%s'",
                     self.message, self.module_name)
        except ImportError as e:
            log.error("Could not import module '%s', exception: %s",
                      self.module_name, e)
            raise
        return None


# NOTE: TO RUN THIS SUCCESSFULLY, YOU NEED TO HAVE THESE
#       PACKAGES INSTALLED:
#       protobuf or python3_protobuf
#       six
#
#       Run the following command to install them:
#       python setup.py install
#
# TO DEBUG: Set DISTUTILS_DEBUG=1 in the environment or run as
# 'python setup.py -vv build_messages'
class build_messages(Command):
    """
    Generates message code mappings. Add to the build process using::

        setup(cmd_class={'build_messages': build_messages})
    """

    description = "generate protocol message code mappings"

    user_options = [
        ('source=', None, 'source CSV file containing message code mappings'),
        ('destination=', None, 'destination Python source file')
    ]

    # Used in loading and generating
    _pb_imports = set()
    _messages = set()
    _linesep = os.linesep
    _indented_item_sep = ',{0}    '.format(_linesep)

    _docstring = [
        ''
        '# This is a generated file. DO NOT EDIT.',
        '',
        '"""',
        'Constants and mappings between Riak protocol codes and messages.',
        '"""',
        ''
    ]

    def initialize_options(self):
        self.source = None
        self.destination = None
        self.update_import = None

    def finalize_options(self):
        if self.source is None:
            self.source = 'riak_pb/src/riak_pb_messages.csv'
        if self.destination is None:
            self.destination = 'riak/pb/messages.py'

    def run(self):
        self.force = True
        self.make_file(self.source, self.destination,
                       self._load_and_generate, [])

    def _load_and_generate(self):
        self._format_python2_or_3()
        self._load()
        self._generate()

    def _load(self):
        with open(self.source, 'r', buffering=1) as csvfile:
            reader = csv.reader(csvfile)
            for row in reader:
                message = MessageCodeMapping(*row)
                self._messages.add(message)
                self._pb_imports.add(message.module_name)

    def _generate(self):
        self._contents = []
        self._generate_doc()
        self._generate_imports()
        self._generate_codes()
        self._generate_classes()
        write_file(self.destination, self._contents)

    def _generate_doc(self):
        # Write the license and docstring header
        self._contents.extend(self._docstring)

    def _generate_imports(self):
        # Write imports
        for im in sorted(self._pb_imports):
            self._contents.append("import {0}".format(im))

    def _generate_codes(self):
        # Write protocol code constants
        self._contents.extend(['', "# Protocol codes"])
        for message in sorted(self._messages):
            self._contents.append("{0} = {1}".format(message.message_code_name,
                                                     message.code))

    def _generate_classes(self):
        # Write message classes
        classes = [self._generate_mapping(message)
                   for message in sorted(self._messages)]

        classes = self._indented_item_sep.join(classes)
        self._contents.extend(['',
                               "# Mapping from code to protobuf class",
                               'MESSAGE_CLASSES = {',
                               '    ' + classes,
                               '}'])

    def _generate_mapping(self, m):
        if m.message_class is not None:
            klass = "{0}.{1}".format(m.module_name,
                                     m.message_class.__name__)
        else:
            klass = "None"
        pair = "{0}: {1}".format(m.message_code_name, klass)
        if len(pair) > 76:
            # Try to satisfy PEP8, lulz
            pair = (self._linesep + '    ').join(pair.split(' '))
        return pair

    def _format_python2_or_3(self):
        """
        Change the PB files to use full pathnames for Python 3.x
        and modify the metaclasses to be version agnostic
        """
        pb_files = set()
        with open(self.source, 'r', buffering=1) as csvfile:
            reader = csv.reader(csvfile)
            for row in reader:
                _, _, proto = row
                pb_files.add('riak/pb/{0}_pb2.py'.format(proto))

        for im in sorted(pb_files):
            with open(im, 'r', buffering=1) as pbfile:
                contents = 'from six import *\n' + pbfile.read()
                contents = re.sub(r'riak_pb2',
                                  r'riak.pb.riak_pb2',
                                  contents)
            # Look for this pattern in the protoc-generated file:
            #
            # class RpbCounterGetResp(_message.Message):
            #    __metaclass__ = _reflection.GeneratedProtocolMessageType
            #
            # and convert it to:
            #
            # @add_metaclass(_reflection.GeneratedProtocolMessageType)
            # class RpbCounterGetResp(_message.Message):
            contents = re.sub(
                r'class\s+(\S+)\((\S+)\):\s*\n'
                '\s+__metaclass__\s+=\s+(\S+)\s*\n',
                r'@add_metaclass(\3)\nclass \1(\2):\n', contents)

            with open(im, 'w', buffering=1) as pbfile:
                pbfile.write(contents)
